src/collapse/dtrace.rs

```diff
@@ -109,21 +151,152 @@ impl Collapse for Folder {
-            occurrences: FnvHashMap::with_capacity_and_hasher(512, fnv::FnvBuildHasher::default()),
-            cache_inlines: Vec::new(),
-            opt,
-            stack_str_size: 0,
-        }
+    #[cfg(test)]
+    fn set_nstacks_per_job(&mut self, n: usize) {
+        self.opt.nstacks_per_job = n;
+    }
+
+    #[cfg(test)]
+    fn set_nthreads(&mut self, n: usize) {
+        self.opt.nthreads = n;
     }
 }
 
 impl Folder {
+    fn collapse_single_threaded<R>(&mut self, mut reader: R) -> io::Result<()>
+    where
+        R: io::BufRead,
+    {
+        let mut line = String::new();
+        loop {
+            line.clear();
+
+            if reader.read_line(&mut line)? == 0 {
+                break;
+            }
+
+            let line = line.trim();
+
+            if line.is_empty() {
+                continue;
+            } else if let Ok(count) = line.parse::<usize>() {
+                self.on_stack_end(count);
+            } else {
+                self.on_stack_line(line);
+            }
+        }
+        Ok(())
+    }
+
+    fn collapse_multi_threaded<R>(&mut self, mut reader: R) -> io::Result<()>
+    where
+        R: io::BufRead,
+    {
+        assert!(self.occurrences.is_concurrent());
+        assert!(self.opt.nstacks_per_job > 0);
+        assert!(self.opt.nthreads > 1);
+
+        enum Message {
+            Job(Vec<u8>),
+            Shutdown,
+        }
+
+        crossbeam::thread::scope(|scope| {
+            // Spin up the threadpool / worker threads.
+            let (tx_input, rx_input) = channel::unbounded();
+            let (tx_output, rx_output) = channel::unbounded();
+            let mut handles = Vec::with_capacity(self.opt.nthreads);
+            for _ in 0..self.opt.nthreads {
+                let rx_input = rx_input.clone();
+                let tx_output = tx_output.clone();
+
+                let occurrences = self.occurrences.clone();
+                let opt = self.opt.clone();
+
+                let handle = scope.spawn(move |_| {
+                    let mut folder = Folder {
+                        cache_inlines: Vec::default(),
+                        occurrences,
+                        stack: VecDeque::default(),
+                        stack_str_size: 0,
+                        opt,
+                    };
+                    loop {
+                        // Receive the data...
+                        let data = match rx_input.recv().unwrap() {
+                            Message::Job(data) => data,
+                            Message::Shutdown => return,
+                        };
+                        // Process the data...
+                        let result = folder.collapse_single_threaded(&data[..]);
+                        // Send the result...
+                        tx_output.send(result).unwrap();
+                        // Prepare for the next round...
+                        folder.stack.clear();
+                        folder.stack_str_size = 0;
```

Comment:
Couldn't we just send the results at the end? Why do we reset + send for every batch we get?

Reply:
Actually, why do we have `tx_output` at all? Nothing meaningful is sent back anyway? Joining all the threads should be sufficient.

Reply:
Oh, this is the whole `Result<T, Box<dyn Any + 'static + Send>>` business, isn't it? Hmm... Okay, fair enough.

My question about why we reset for each batch still stands though.

Reply:
I don't think this has quite been resolved. Why do we have to reset the folder for each iteration?

Reply:
`collapse_single_threaded` doesn't clear the stack buffer. Don't we need to clear the stack buffer when we start a new stack?

Reply:
oh, nevermind, it does.

Reply:
Actually, imagine `collapse_single_threaded` (shown below) is doing it's thing and, all of a sudden, we reach EOF but the last line processed did not take the `self.on_stack_end` branch below, but rather it took the `self.on_stack_line` branch. In that case, you might have internal state that should have been, but was never reset, since the resetting of the state happens in `self.on_stack_end`.

Now, you might say that if this occurs the input data is malformed and perhaps we should error, but right now I don't see that we do anything if the file finishes with a malformed stack (i.e. one that doesn't have the correct ending).

What would you like to do about this? My reaction was - I'm not trying to solve this problem; so let's just double check that we can never start processing a new stack with stale internal state sitting round - hence the clears, etc. But perhaps we should think about this problem instead of just doing what I did.

```
fn collapse_single_threaded<R>(&mut self, mut reader: R) -> io::Result<()>
    where
        R: io::BufRead,
    {
        let mut line = String::new();
        loop {
            line.clear();

            if reader.read_line(&mut line)? == 0 {
                break;
            }

            let line = line.trim();

            if line.is_empty() {
                continue;
            } else if let Ok(count) = line.parse::<usize>() {
                self.on_stack_end(count);
            } else {
                self.on_stack_line(line);
            }
        }
        Ok(())
    }
```

Reply:
I guess that can't happen in the multi-threaded code because we ensure that we're sending properly formed stacks to the children. But should we do about this in the single-threaded case? Shouldn't we return an error if the user provides data that ends like, e.g., this...

```
CPU     ID                    FUNCTION:NAME
  8  35077                        :tick-60s

              0x104d08831
              0x104cecffd
              0x104cecffd
              0x104cecffd
              0x104ce54e7
              libjvm.dylib`JavaCalls::call_helper(JavaValue*, methodHandle*, JavaCallArguments*, Thread*)+0x6ae
              libjvm.dylib`JavaCalls::call_virtual(JavaValue*, KlassHandle, Symbol*, Symbol*, JavaCallArguments*, Thread*)+0x164
              libjvm.dylib`JavaCalls::call_virtual(JavaValue*, Handle, KlassHandle, Symbol*, Symbol*, Thread*)+0x4a
              libjvm.dylib`thread_entry(JavaThread*, Thread*)+0x7c
              libjvm.dylib`JavaThread::thread_main_inner()+0x9b
              libjvm.dylib`JavaThread::run()+0x1c0
              libjvm.dylib`java_start(Thread*)+0xf6
              libsystem_pthread.dylib`_pthread_body+0x7e
              libsystem_pthread.dylib`_pthread_start+0x46
              libsystem_pthread.dylib`thread_start+0xd
                1

              libjvm.dylib`Node::replace_edge(Node*, Node*)+0x4f
              libjvm.dylib`PhaseIterGVN::remove_globally_dead_node(Node*)+0x109
              libjvm.dylib`PhaseIterGVN::subsume_node(Node*, Node*)+0x199
```

Reply:
I think we should error out in `if reader.read_line(&mut line)? == 0 {` (EOF) if we detect that we're in this case :)